### ISSUE TITLE
Optional task input property annotation ignored

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#11840]
@JLLeitschuh reported a problem that can be reproduced by:

1. Checkout the branch https://github.com/gradle/dotcom-docs/tree/chore/JLL/update_to_gradle_6.0.1
2. Delete the = MARKER from this line: https://github.com/gradle/dotcom-docs/pull/100/files#diff-9ea661cc32820ffb1b224c5a1ffe19dfR18
3. Restore this line to what it was before: https://github.com/gradle/dotcom-docs/pull/100/files#diff-9ea661cc32820ffb1b224c5a1ffe19dfL29
4. Run any Gradle task that causes `:buildSrc` to get executed

The problems are reported in test failures, saying the `@Optional` property `sitemapUrl` in `RobotsGatheringDuplicateContentHandler` has no value specified.

---
cc: @gradle/build-cache

**Comment by lptr** [comment]
I'm not sure what's happening here. When I run `RobotsTxtGeneratingIntegrationTest .duplicate robots.txt handler works as expected`, I get this error indeed after making the changes @JLLeitschuh recommends:

```text
Output:
> Task :fetchFirstSourceSource
> Task :transformFirstSourceSourceForTestPublicationPublication
> Task :fetchOtherwiseExcludedSource
> Task :transformOtherwiseExcludedSourceForTestPublicationPublication
> Task :fetchSecondSourceSource
> Task :transformSecondSourceSourceForTestPublicationPublication
> Task :handleDuplicatesForTestPublicationPublication FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':handleDuplicatesForTestPublicationPublication' (type 'DuplicateContentHandlerTask').
> No value has been specified for property 'sitemapUrl'.
```

However, I'm not sure how we get to this error. The task of type `DuplicateContentHandlerTask` itself should not have a property called `sitemapUrl`, as that property is only defined in `RobotsGatheringDuplicateContentHandler`. I wonder why we don't display the fully-qualified property name here. That alone should be a bug I think.

But I don't understand how that property appears here nevertheless. The task has only the following properties:

```groovy
    @OutputFile
    File manifestFile

    @OutputDirectory
    File outputDir

    @Internal
    SourcesDocumentationPublication publication

    @InputFiles
    @PathSensitive(value = PathSensitivity.RELATIVE)
    protected FileCollection getCompositeSources() {
        return publication.compositeSources
    }
```

None of these expose an instance of `RobotsGatheringDuplicateContentHandler` anywhere. I'll guess I have to conjure up a debugger after all... :/

**Comment by lptr** [comment]
OK, so there's something reeeeeeally fishy going on here. A hint about the presence of `InputMapper` would have saved a lot of time for me. :) Apparently @mark-vieira added a workaround for a limitation where child properties of `@Nested` input properties were not tracked properly wrt polymorphic values. This is stated in the comments at the top of the class. (Thanks Mark for being thorough!)

The limitation has soon been lifter after this workaround was made, so `InputMapper` is not needed anymore, and `@Nested` should be used instead.

To explain the phenomena you've experienced: Mark's workaround simply re-registers the nested properties (without capturing their nested names). His code does not handle `@Optional` (I think probably because at that time we didn't check `@Input` values for `null`s, which is a whole another story).

Anyway, just drop this class and use `@Nested`, and it will work as you'd expect it.

**Comment by lptr** [comment]
Rich (=polymorphic) `@Nested` inputs have been supported since Gradle 4.5: https://github.com/gradle/gradle/pull/3720.

**Comment by mark-vieira** [comment]
> Rich (=polymorphic) @Nested inputs have been supported since Gradle 4.5: #3720.

Yes, and that would make the `InputMapper` unnecessary, however the project you are referring to (the docs generator) is still on Gradle 3.5 because it relies heavily on the `@ParallelizableTask` annotation to be able to run all these tasks in parallel despite being part of a single project. Migrating to Gradle 4.0+ would mean moving to using the worker API which when I investigated it at the time was non-trivial due to the reliance on non-serializable types like `FileTree` on many of these task inputs. It _could_ be done, but it'd be a big job. Since that build is stuck (so far as I know) on Gradle 3.5, fixing `InputMapper` to handle this case is probably the simpler bet.

**Comment by lptr** [comment]
Hey @mark-vieira :) IIUC, the branch we are talking about has already been upgraded to Gradle 6.0.1, so that shouldn't be a problem I think.

**Comment by mark-vieira** [comment]
Awesome. Then yes, my suggestion is to completely remove `InputMapper` since `@Nested` in Gradle 6.0 behaves they way it should.
